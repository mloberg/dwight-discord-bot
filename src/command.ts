import { SlashCommandBuilder, SlashCommandIntegerOption } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';

type Handler = (command: CommandInteraction) => Promise<void>;

export class CommandBuilder extends SlashCommandBuilder {
    constructor(readonly handler: Handler) {
        super();
    }
}

export const rollOption = (dice: string): SlashCommandIntegerOption =>
    new SlashCommandIntegerOption().setName('roll').setDescription(`Value of a ${dice} roll.`);
