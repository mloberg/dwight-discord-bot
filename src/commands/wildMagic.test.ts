import { Client, Guild, Message, TextChannel } from 'discord.js';

import command, { wildMagic } from './wildMagic';

const mocks = {
    delete: jest.fn(),
    reply: jest.fn(),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Collection: jest.fn(),
    Message: jest.fn().mockImplementation(() => {
        return {
            delete: mocks.delete,
            reply: mocks.reply,
        };
    }),
}));

describe('_wild configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('wild');
        expect(command.description).toEqual('Roll on the Wild Magic table');
        expect(command.usage).toEqual('[ROLL]');
    });

    it('should have aliases', () => {
        expect(command.alias).toContain('wild-magic');
    });
});

describe('_wild', () => {
    let message: Message;

    beforeEach(() => {
        mocks.delete.mockClear();
        mocks.reply.mockClear();
        mocks.reply.mockReturnThis();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it('returns a wild magic event', async () => {
        const reply = await command.run(message, { $0: 'wild', _: [] });

        expect(mocks.delete).toHaveBeenCalledTimes(1);
        expect(reply).toBe(message);

        const event = mocks.reply.mock.calls[0][0];
        expect(wildMagic).toContainEqual(event);
    });

    it('returns a wild magic event for a dice roll', async () => {
        const one = await command.run(message, { $0: 'wild', _: ['89'] });
        const two = await command.run(message, { $0: 'wild', _: [], roll: 97 });

        expect(mocks.delete).toHaveBeenCalledTimes(2);
        expect(one).toBe(message);
        expect(two).toBe(message);

        expect(mocks.reply).toHaveBeenCalledWith(
            'You become invisible for the next minute. During that time, other creatures canâ€™t hear you. The invisibility ends if you attack or cast a spell.',
        );
        expect(mocks.reply).toHaveBeenCalledWith('You are surrounded by faint, ethereal music for the next minute.');
    });
});
