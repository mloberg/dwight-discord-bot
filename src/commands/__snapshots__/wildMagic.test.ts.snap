// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/wildMagic is a slash command 1`] = `
Object {
  "default_permission": undefined,
  "description": "Roll on the Wild Magic table",
  "name": "wildmagic",
  "options": Array [
    Object {
      "description": "Roll on the Barbarian Wild Magic table",
      "name": "barbarian",
      "required": false,
      "type": 5,
    },
    Object {
      "choices": undefined,
      "description": "Value of a d100 (Sorcerer) or d8 (Barbarian) roll.",
      "name": "roll",
      "required": false,
      "type": 4,
    },
  ],
}
`;
